File: src/Client/CineMol/Styles.fs
module Client.CineMol.Styles


type RGB = int * int * int
type Alpha = float
type DiffusionGradient = Alpha[]

type Color = { RGB: RGB; Alpha: Alpha }
    with
    member x.Diffuse (alpha : float) : float * float * float =
        let R, G, B = x.RGB
        (float R) * alpha, (float G) * alpha, (float B * alpha)

let atomColorGradient: DiffusionGradient = [| 1.0; 0.89; 0.66; 0.34; 0.0 |]

type AtomType =
    | H  | He | Li | Be | B  | C  | N  | O  | F  | Ne | Na | Mg | Al | Si
    | P  | S  | Cl | Ar | K  | Ca | Sc | Ti | V  | Cr | Mn | Fe | Co | Ni
    | Cu | Zn | Ga | Ge | As | Se | Br | Kr | Rb | Sr | Y  | Zr | Nb | Mo
    | Tc | Ru | Rh | Pd | Ag | Cd | In | Sn | Sb | Te | I  | Xe | Cs | Ba
    | La | Ce | Pr | Nd | Pm | Sm | Eu | Gd | Tb | Dy | Ho | Er | Tm | Yb
    | Lu | Hf | Ta | W  | Re | Os | Ir | Pt | Au | Hg | Tl | Pb | Bi | Po
    | At | Rn | Fr | Ra | Ac | Th | Pa | U  | Np | Pu | Am | Cm | Bk | Cf
    | Es | Fm | Md | No | Lr | Rf | Db | Sg | Bh | Hs | Mt | Ds | Rg | Cn
    | Nh | Fl | Mc | Lv | Ts | Og | Unknown

type AtomColorStyle = | CPK

let getAtomColor (style: AtomColorStyle) (atomType: AtomType) : Color =
    match style with
    | CPK ->
        match atomType with
        | H  -> 255, 255, 255 | He -> 255, 255, 192 | Li -> 204, 128, 255
        | Be -> 194, 255, 000 | B  -> 255, 181, 181 | C  -> 144, 144, 144
        | N  -> 048, 080, 248 | O  -> 255, 013, 013 | F  -> 144, 224, 080
        | Ne -> 179, 227, 245 | Na -> 171, 092, 242 | Mg -> 138, 255, 000
        | Al -> 191, 166, 166 | Si -> 240, 200, 160 | P  -> 255, 128, 000
        | S  -> 255, 255, 048 | Cl -> 031, 240, 031 | Ar -> 128, 209, 227
        | K  -> 143, 064, 212 | Ca -> 061, 255, 000 | Sc -> 230, 230, 230
        | Ti -> 191, 194, 199 | V  -> 166, 166, 171 | Cr -> 138, 153, 199
        | Mn -> 156, 122, 199 | Fe -> 224, 102, 051 | Co -> 240, 144, 160
        | Ni -> 080, 208, 080 | Cu -> 200, 128, 051 | Zn -> 125, 128, 176
        | Ga -> 194, 143, 143 | Ge -> 102, 143, 143 | As -> 189, 128, 227
        | Se -> 255, 161, 000 | Br -> 166, 041, 041 | Kr -> 092, 184, 209
        | Rb -> 112, 046, 176 | Sr -> 000, 255, 000 | Y  -> 148, 255, 255
        | Zr -> 148, 224, 224 | Nb -> 115, 194, 201 | Mo -> 084, 181, 181
        | Tc -> 059, 158, 158 | Ru -> 036, 143, 143 | Rh -> 010, 125, 140
        | Pd -> 000, 105, 133 | Ag -> 192, 192, 192 | Cd -> 255, 217, 143
        | In -> 166, 117, 115 | Sn -> 102, 128, 128 | Sb -> 158, 099, 181
        | Te -> 212, 122, 000 | I  -> 148, 000, 148 | Xe -> 066, 158, 176
        | Cs -> 087, 023, 143 | Ba -> 000, 201, 000 | La -> 112, 212, 255
        | Ce -> 255, 255, 199 | Pr -> 217, 255, 199 | Nd -> 199, 255, 199
        | Pm -> 163, 255, 199 | Sm -> 143, 255, 199 | Eu -> 097, 255, 199
        | Gd -> 069, 255, 199 | Tb -> 048, 255, 199 | Dy -> 031, 255, 199
        | Ho -> 000, 255, 156 | Er -> 000, 230, 117 | Tm -> 000, 212, 082
        | Yb -> 000, 191, 056 | Lu -> 000, 171, 036 | Hf -> 077, 194, 255
        | Ta -> 077, 166, 255 | W  -> 033, 148, 214 | Re -> 038, 125, 171
        | Os -> 038, 102, 150 | Ir -> 023, 084, 135 | Pt -> 208, 208, 224
        | Au -> 255, 209, 035 | Hg -> 184, 184, 208 | Tl -> 166, 084, 077
        | Pb -> 087, 089, 097 | Bi -> 158, 079, 181 | Po -> 171, 092, 000
        | At -> 117, 079, 069 | Rn -> 066, 130, 150 | Fr -> 066, 000, 102
        | Ra -> 000, 125, 000 | Ac -> 112, 171, 250 | Th -> 000, 186, 255
        | Pa -> 000, 161, 255 | U  -> 000, 143, 255 | Np -> 000, 128, 255
        | Pu -> 000, 107, 255 | Am -> 084, 092, 242 | Cm -> 120, 092, 227
        | Bk -> 138, 079, 227 | Cf -> 161, 054, 212 | Es -> 179, 031, 212
        | Fm -> 179, 031, 186 | Md -> 179, 013, 166 | No -> 189, 013, 135
        | Lr -> 199, 000, 102 | Rf -> 204, 000, 089 | Db -> 209, 000, 079
        | Sg -> 217, 000, 069 | Bh -> 224, 000, 056 | Hs -> 230, 000, 046
        | Mt -> 235, 000, 038 | _  -> 000, 000, 000
        |> (fun rgb -> { RGB = rgb; Alpha = 1.0 })

type AtomGeomStyle = | Default
type Radius = float

let getAtomRadius (style: AtomGeomStyle) (atomType: AtomType) : Radius =
    match style with
    | Default ->
        // Atom radii in pico meters (based on quantum mechanical wave functions).
        match atomType with
        | H  -> 053 | He -> 031 | Li -> 167 | Be -> 112 | B  -> 087 | C  -> 067
        | N  -> 056 | O  -> 048 | F  -> 042 | Ne -> 038 | Na -> 190 | Mg -> 145
        | Al -> 118 | Si -> 111 | P  -> 098 | S  -> 088 | Cl -> 079 | Ar -> 071
        | K  -> 243 | Ca -> 194 | Sc -> 184 | Ti -> 176 | V  -> 171 | Cr -> 166
        | Mn -> 161 | Fe -> 156 | Co -> 152 | Ni -> 149 | Cu -> 145 | Zn -> 142
        | Ga -> 136 | Ge -> 125 | As -> 114 | Se -> 103 | Br -> 094 | Kr -> 088
        | Rb -> 265 | Sr -> 219 | Y  -> 212 | Zr -> 206 | Nb -> 198 | Mo -> 190
        | Tc -> 183 | Ru -> 178 | Rh -> 173 | Pd -> 169 | Ag -> 165 | Cd -> 155
        | In -> 156 | Sn -> 145 | Sb -> 133 | Te -> 123 | I  -> 115 | Xe -> 108
        | Cs -> 298 | Ba -> 253 | Hf -> 208 | Ta -> 200 | W  -> 193 | Re -> 188
        | Os -> 185 | Ir -> 180 | Pt -> 177 | Au -> 174 | Hg -> 171 | Tl -> 156
        | Pb -> 154 | Bi -> 143 | Po -> 135 | At -> 127 | Rn -> 120 | _  -> 031

let normalizeRadius (style: AtomGeomStyle) (radius: Radius) : Radius =
    match style with
    // Default atom radius is normalized by dividing by hydrogen atom radius.
    | Default -> radius / (getAtomRadius Default AtomType.H)

File: cinemol-py/styles.py
from __future__ import annotations
from array import array
from dataclasses import dataclass
from typing import (Tuple, Any, List)
from .fable_modules.fable_library.double import divide
from .fable_modules.fable_library.reflection import (TypeInfo, int32_type, tuple_type, float64_type, record_type, union_type)
from .fable_modules.fable_library.types import (Record, Array, Union)

def _expr1() -> TypeInfo:
    return record_type("Client.CineMol.Styles.Color", [], Color, lambda: [("RGB", tuple_type(int32_type, int32_type, int32_type)), ("Alpha", float64_type)])


@dataclass(eq = False, repr = False)
class Color(Record):
    RGB: Tuple[int, int, int]
    Alpha: float

Color_reflection = _expr1

def Color__Diffuse_5E38073B(x: Color, alpha: float) -> Tuple[float, float, float]:
    pattern_input: Tuple[int, int, int] = x.RGB
    return (pattern_input[0] * alpha, pattern_input[1] * alpha, pattern_input[2] * alpha)


atom_color_gradient: Array[float] = array("d", [1.0, 0.89, 0.66, 0.34, 0.0])

def _expr3() -> TypeInfo:
    return union_type("Client.CineMol.Styles.AtomType", [], AtomType, lambda: [[], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], []])


class AtomType(Union):
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> List[str]:
        return ["H", "He", "Li", "Be", "B", "C", "N", "O", "F", "Ne", "Na", "Mg", "Al", "Si", "P", "S", "Cl", "Ar", "K", "Ca", "Sc", "Ti", "V", "Cr", "Mn", "Fe", "Co", "Ni", "Cu", "Zn", "Ga", "Ge", "As", "Se", "Br", "Kr", "Rb", "Sr", "Y", "Zr", "Nb", "Mo", "Tc", "Ru", "Rh", "Pd", "Ag", "Cd", "In", "Sn", "Sb", "Te", "I", "Xe", "Cs", "Ba", "La", "Ce", "Pr", "Nd", "Pm", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho", "Er", "Tm", "Yb", "Lu", "Hf", "Ta", "W", "Re", "Os", "Ir", "Pt", "Au", "Hg", "Tl", "Pb", "Bi", "Po", "At", "Rn", "Fr", "Ra", "Ac", "Th", "Pa", "U", "Np", "Pu", "Am", "Cm", "Bk", "Cf", "Es", "Fm", "Md", "No", "Lr", "Rf", "Db", "Sg", "Bh", "Hs", "Mt", "Ds", "Rg", "Cn", "Nh", "Fl", "Mc", "Lv", "Ts", "Og", "Unknown"]


AtomType_reflection = _expr3

def _expr4() -> TypeInfo:
    return union_type("Client.CineMol.Styles.AtomColorStyle", [], AtomColorStyle, lambda: [[]])


class AtomColorStyle(Union):
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> List[str]:
        return ["CPK"]


AtomColorStyle_reflection = _expr4

def get_atom_color(style: AtomColorStyle, atom_type: AtomType) -> Color:
    return Color(((255, 255, 255)) if (atom_type.tag == 0) else (((255, 255, 192)) if (atom_type.tag == 1) else (((204, 128, 255)) if (atom_type.tag == 2) else (((194, 255, 0)) if (atom_type.tag == 3) else (((255, 181, 181)) if (atom_type.tag == 4) else (((144, 144, 144)) if (atom_type.tag == 5) else (((48, 80, 248)) if (atom_type.tag == 6) else (((255, 13, 13)) if (atom_type.tag == 7) else (((144, 224, 80)) if (atom_type.tag == 8) else (((179, 227, 245)) if (atom_type.tag == 9) else (((171, 92, 242)) if (atom_type.tag == 10) else (((138, 255, 0)) if (atom_type.tag == 11) else (((191, 166, 166)) if (atom_type.tag == 12) else (((240, 200, 160)) if (atom_type.tag == 13) else (((255, 128, 0)) if (atom_type.tag == 14) else (((255, 255, 48)) if (atom_type.tag == 15) else (((31, 240, 31)) if (atom_type.tag == 16) else (((128, 209, 227)) if (atom_type.tag == 17) else (((143, 64, 212)) if (atom_type.tag == 18) else (((61, 255, 0)) if (atom_type.tag == 19) else (((230, 230, 230)) if (atom_type.tag == 20) else (((191, 194, 199)) if (atom_type.tag == 21) else (((166, 166, 171)) if (atom_type.tag == 22) else (((138, 153, 199)) if (atom_type.tag == 23) else (((156, 122, 199)) if (atom_type.tag == 24) else (((224, 102, 51)) if (atom_type.tag == 25) else (((240, 144, 160)) if (atom_type.tag == 26) else (((80, 208, 80)) if (atom_type.tag == 27) else (((200, 128, 51)) if (atom_type.tag == 28) else (((125, 128, 176)) if (atom_type.tag == 29) else (((194, 143, 143)) if (atom_type.tag == 30) else (((102, 143, 143)) if (atom_type.tag == 31) else (((189, 128, 227)) if (atom_type.tag == 32) else (((255, 161, 0)) if (atom_type.tag == 33) else (((166, 41, 41)) if (atom_type.tag == 34) else (((92, 184, 209)) if (atom_type.tag == 35) else (((112, 46, 176)) if (atom_type.tag == 36) else (((0, 255, 0)) if (atom_type.tag == 37) else (((148, 255, 255)) if (atom_type.tag == 38) else (((148, 224, 224)) if (atom_type.tag == 39) else (((115, 194, 201)) if (atom_type.tag == 40) else (((84, 181, 181)) if (atom_type.tag == 41) else (((59, 158, 158)) if (atom_type.tag == 42) else (((36, 143, 143)) if (atom_type.tag == 43) else (((10, 125, 140)) if (atom_type.tag == 44) else (((0, 105, 133)) if (atom_type.tag == 45) else (((192, 192, 192)) if (atom_type.tag == 46) else (((255, 217, 143)) if (atom_type.tag == 47) else (((166, 117, 115)) if (atom_type.tag == 48) else (((102, 128, 128)) if (atom_type.tag == 49) else (((158, 99, 181)) if (atom_type.tag == 50) else (((212, 122, 0)) if (atom_type.tag == 51) else (((148, 0, 148)) if (atom_type.tag == 52) else (((66, 158, 176)) if (atom_type.tag == 53) else (((87, 23, 143)) if (atom_type.tag == 54) else (((0, 201, 0)) if (atom_type.tag == 55) else (((112, 212, 255)) if (atom_type.tag == 56) else (((255, 255, 199)) if (atom_type.tag == 57) else (((217, 255, 199)) if (atom_type.tag == 58) else (((199, 255, 199)) if (atom_type.tag == 59) else (((163, 255, 199)) if (atom_type.tag == 60) else (((143, 255, 199)) if (atom_type.tag == 61) else (((97, 255, 199)) if (atom_type.tag == 62) else (((69, 255, 199)) if (atom_type.tag == 63) else (((48, 255, 199)) if (atom_type.tag == 64) else (((31, 255, 199)) if (atom_type.tag == 65) else (((0, 255, 156)) if (atom_type.tag == 66) else (((0, 230, 117)) if (atom_type.tag == 67) else (((0, 212, 82)) if (atom_type.tag == 68) else (((0, 191, 56)) if (atom_type.tag == 69) else (((0, 171, 36)) if (atom_type.tag == 70) else (((77, 194, 255)) if (atom_type.tag == 71) else (((77, 166, 255)) if (atom_type.tag == 72) else (((33, 148, 214)) if (atom_type.tag == 73) else (((38, 125, 171)) if (atom_type.tag == 74) else (((38, 102, 150)) if (atom_type.tag == 75) else (((23, 84, 135)) if (atom_type.tag == 76) else (((208, 208, 224)) if (atom_type.tag == 77) else (((255, 209, 35)) if (atom_type.tag == 78) else (((184, 184, 208)) if (atom_type.tag == 79) else (((166, 84, 77)) if (atom_type.tag == 80) else (((87, 89, 97)) if (atom_type.tag == 81) else (((158, 79, 181)) if (atom_type.tag == 82) else (((171, 92, 0)) if (atom_type.tag == 83) else (((117, 79, 69)) if (atom_type.tag == 84) else (((66, 130, 150)) if (atom_type.tag == 85) else (((66, 0, 102)) if (atom_type.tag == 86) else (((0, 125, 0)) if (atom_type.tag == 87) else (((112, 171, 250)) if (atom_type.tag == 88) else (((0, 186, 255)) if (atom_type.tag == 89) else (((0, 161, 255)) if (atom_type.tag == 90) else (((0, 143, 255)) if (atom_type.tag == 91) else (((0, 128, 255)) if (atom_type.tag == 92) else (((0, 107, 255)) if (atom_type.tag == 93) else (((84, 92, 242)) if (atom_type.tag == 94) else (((120, 92, 227)) if (atom_type.tag == 95) else (((138, 79, 227)) if (atom_type.tag == 96) else (((161, 54, 212)) if (atom_type.tag == 97) else (((179, 31, 212)) if (atom_type.tag == 98) else (((179, 31, 186)) if (atom_type.tag == 99) else (((179, 13, 166)) if (atom_type.tag == 100) else (((189, 13, 135)) if (atom_type.tag == 101) else (((199, 0, 102)) if (atom_type.tag == 102) else (((204, 0, 89)) if (atom_type.tag == 103) else (((209, 0, 79)) if (atom_type.tag == 104) else (((217, 0, 69)) if (atom_type.tag == 105) else (((224, 0, 56)) if (atom_type.tag == 106) else (((230, 0, 46)) if (atom_type.tag == 107) else (((235, 0, 38)) if (atom_type.tag == 108) else ((0, 0, 0)))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))), 1.0)


def _expr5() -> TypeInfo:
    return union_type("Client.CineMol.Styles.AtomGeomStyle", [], AtomGeomStyle, lambda: [[]])


class AtomGeomStyle(Union):
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> List[str]:
        return ["Default"]


AtomGeomStyle_reflection = _expr5

def get_atom_radius(style: AtomGeomStyle, atom_type: AtomType) -> float:
    return 37 if (atom_type.tag == 0) else (152 if (atom_type.tag == 2) else (112 if (atom_type.tag == 3) else (88 if (atom_type.tag == 4) else (77 if (atom_type.tag == 5) else (70 if (atom_type.tag == 6) else (66 if (atom_type.tag == 7) else (64 if (atom_type.tag == 8) else (186 if (atom_type.tag == 10) else (160 if (atom_type.tag == 11) else (143 if (atom_type.tag == 12) else (117 if (atom_type.tag == 13) else (110 if (atom_type.tag == 14) else (104 if (atom_type.tag == 15) else (99 if (atom_type.tag == 16) else (231 if (atom_type.tag == 18) else (197 if (atom_type.tag == 19) else (160 if (atom_type.tag == 20) else (146 if (atom_type.tag == 21) else (131 if (atom_type.tag == 22) else (125 if (atom_type.tag == 23) else (129 if (atom_type.tag == 24) else (126 if (atom_type.tag == 25) else (126 if (atom_type.tag == 26) else (124 if (atom_type.tag == 27) else (128 if (atom_type.tag == 28) else (133 if (atom_type.tag == 29) else (122 if (atom_type.tag == 30) else (122 if (atom_type.tag == 31) else (121 if (atom_type.tag == 32) else (117 if (atom_type.tag == 33) else (114 if (atom_type.tag == 34) else (241 if (atom_type.tag == 36) else (215 if (atom_type.tag == 37) else (180 if (atom_type.tag == 38) else (157 if (atom_type.tag == 39) else (143 if (atom_type.tag == 40) else (136 if (atom_type.tag == 41) else (130 if (atom_type.tag == 42) else (133 if (atom_type.tag == 43) else (134 if (atom_type.tag == 44) else (138 if (atom_type.tag == 45) else (144 if (atom_type.tag == 46) else (149 if (atom_type.tag == 47) else (162 if (atom_type.tag == 48) else (140 if (atom_type.tag == 49) else (141 if (atom_type.tag == 50) else (137 if (atom_type.tag == 51) else (133 if (atom_type.tag == 52) else (262 if (atom_type.tag == 54) else (217 if (atom_type.tag == 55) else (157 if (atom_type.tag == 71) else (143 if (atom_type.tag == 72) else (137 if (atom_type.tag == 73) else (137 if (atom_type.tag == 74) else (134 if (atom_type.tag == 75) else (135 if (atom_type.tag == 76) else (138 if (atom_type.tag == 77) else (144 if (atom_type.tag == 78) else (150 if (atom_type.tag == 79) else (171 if (atom_type.tag == 80) else (175 if (atom_type.tag == 81) else (146 if (atom_type.tag == 82) else (140 if (atom_type.tag == 83) else (140 if (atom_type.tag == 84) else 77))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))


def normalize_radius(style: AtomGeomStyle, radius: float) -> float:
    return divide(radius, get_atom_radius(AtomGeomStyle(0), AtomType(5)))


__all__ = ["Color_reflection", "Color__Diffuse_5E38073B", "atom_color_gradient", "AtomType_reflection", "AtomColorStyle_reflection", "get_atom_color", "AtomGeomStyle_reflection", "get_atom_radius", "normalize_radius"]

